src/audio/module_adapter/module/volume/volume_generic.c

```diff
@@ -233,8 +221,181 @@ static void vol_s16_to_s16(struct processing_module *mod, struct input_stream_bu
 	/* update peak vol */
 	peak_vol_update(cd);
 }
+#endif
+
+#else /* functions without peak volume detection*/
+
+#if CONFIG_FORMAT_S24LE
+/**
+ * \brief Volume s24 to s24 multiply function
+ * \param[in] x   input sample.
+ * \param[in] vol gain.
+ * \return output sample.
+ *
+ * Volume multiply for 24 bit input and 24 bit output.
+ */
+static inline int32_t vol_mult_s24_to_s24(int32_t x, int32_t vol)
+{
+	return q_multsr_sat_32x32_24(sign_extend_s24(x), vol, Q_SHIFT_BITS_64(23, VOL_QXY_Y, 23));
+}
+
+/**
+ * \brief Volume processing from 24/32 bit to 24/32 bit.
+ * \param[in,out] dev Volume base component device.
+ * \param[in,out] sink Destination buffer.
+ * \param[in,out] source Input buffer.
+ * \param[in] frames Number of frames to process.
+ *
+ * Copy and scale volume from 24/32 bit source buffer
+ * to 24/32 bit destination buffer.
+ */
+static void vol_s24_to_s24(struct processing_module *mod, struct input_stream_buffer *bsource,
+			   struct output_stream_buffer *bsink, uint32_t frames)
+{
+	struct vol_data *cd = module_get_private_data(mod);
+	struct audio_stream __sparse_cache *source = bsource->data;
+	struct audio_stream __sparse_cache *sink = bsink->data;
+	int32_t vol;
+	int32_t *x, *x0;
+	int32_t *y, *y0;
+	int nmax, n, i, j;
+	const int nch = source->channels;
+	int remaining_samples = frames * nch;
+
+	x = source->r_ptr;
+	y = sink->w_ptr;
+
+	bsource->consumed += VOL_S32_SAMPLES_TO_BYTES(remaining_samples);
+	bsink->size += VOL_S32_SAMPLES_TO_BYTES(remaining_samples);
+	while (remaining_samples) {
+		nmax = audio_stream_samples_without_wrap_s24(source, x);
+		n = MIN(remaining_samples, nmax);
+		nmax = audio_stream_samples_without_wrap_s24(sink, y);
+		n = MIN(n, nmax);
+		for (j = 0; j < nch; j++) {
+			x0 = x + j;
+			y0 = y + j;
+			vol = cd->volume[j];
+			for (i = 0; i < n; i += nch) {
+				*y0 = vol_mult_s24_to_s24(*x0, vol);
+				x0 += nch;
+				y0 += nch;
+			}
```

Comment:
I think this should be easier and use just as many cycles or potentially fewer:
```
for (j = 0; j < nch; j++)
	y0[j] = vol_mult_s24_to_s24(x0[j], vol);
```
Also in similar other cases above and below

Reply:
thanks